### PR TITLE
[WIP] Frontend support for creation of snapshots - Part 2

### PR DESCRIPTION
**Pull request by pavannd1**
As mentioned in #221, this PR will add changes to watch for creation of a VolumeSnapshot CR and trigger the creation of snapshot. Once a snapshot is created, the VolumeSnapshotStatus is updated with relevant information.
A follow up PR would have changes to watch for PVC creation with the above created snapshot as the source.

### CODE REVIEW

File: frontend/kubernetes/plugin.go [hunk 1]
```diff
@@ -1508,3 +1543,191 @@ func (p *Plugin) updatePVCSize(
...
+			"volumeSnapshot":  volSnap.Name,
+			"sourcePVC":       claim.Name,
+			"sourcePVCStatus": claim.Status.Phase,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not " +
+			"bound to any volume")
+		return
+	}
+
+	// Create the volume configuration object
+	uniqueClaimName := getUniqueClaimName(claim)
```
Comment (clintonk) [inline frontend/kubernetes/plugin.go]: I don't understand why we need to call getUniqueClaimName here.  We need to pass the name of an existing volume (i.e., the _PV_ that the source PVC is bound to) orchestrator.CreateVolumeSnapshot().

File: frontend/kubernetes/plugin.go [hunk 2]
```diff
@@ -1508,3 +1543,191 @@ func (p *Plugin) updatePVCSize(
...
+
+	// Create the snapshot
+	snapshotExt, err := p.orchestrator.CreateVolumeSnapshot(volSnap.Name, uniqueClaimName)
```
Comment (clintonk) [inline frontend/kubernetes/plugin.go]: I must be missing something fundamental here.  volSnap is a VolumeSnapshot object, which is a namespaced, user-visible object, but we are persisting it here in Trident (which manages objects across all namespaces), implying that volSnap.Name is unique in the cluster.  Is that true?

And where is the code that creates VolumeSnapshotContent objects?  My expectation is that Trident would notice a new VolumeSnapshot and react by creating a new VolumeSnapshotContent object (whose name Trident would create and could guarantee would be unique in the cluster).  That would be consistent with how Trident notices new PVCs and creates associated PVs.

File: frontend/kubernetes/plugin.go (hunk 1 above)
Comment (pavannd1) [reply frontend/kubernetes/plugin.go]: You are right. This should be the PV name. Will update this.

File: frontend/kubernetes/plugin.go (hunk 2 above)
Comment (clintonk) [reply frontend/kubernetes/plugin.go]: Hi, @pavannd1, did you have any thoughts on my comments here?  I'm not sure we have a workable approach here in Part 2, which is why I've held off merging the Part 1 patch.

Comment (pavannd1) [reply frontend/kubernetes/plugin.go]: @clintonk Sorry for the delayed response. Yes, these are some good points and I have been trying out a couple different solutions for this. I will get back to you with the details about my tests soon.

**Comment by pavannd1** [comment]
@clintonk Here are the details

### Test Setup

- Kubernetes cluster on AWS
- Trident deployed with AWS-CVS as backend
- CSI `hostpath` driver along with `csi-provisioner` and `csi-snapshotter`

### Snapshot Flow

**User**
- Create a custom `VolumeSnapshotClass` with `snapshotter` as Trident. This is not registered with any driver and can be considered as a dummy class.
```
apiVersion: snapshot.storage.k8s.io/v1alpha1
kind: VolumeSnapshotClass
metadata:
  name: trident-snapclass
  annotations:
    snapshot.storage.kubernetes.io/is-default-class: "true"
snapshotter: netapp.io/trident
```
- Create a trident PVC.
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: podpvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: basic
```
- Create a `VolumeSnapshot` for the trident PVC with the dummy snapshot class.
```
apiVersion: snapshot.storage.k8s.io/v1alpha1
kind: VolumeSnapshot
metadata:
  name: snapshot-test
spec:
  snapshotClassName: trident-snapclass
  source:
    name: podpvc
    kind: PersistentVolumeClaim
```
This `VolumeSnapshot` is ignored by the `csi-snapshotter` since the class is not associated with the registered driver. (Currently, each external-snapshotter can be configured with only a single driver)

**Trident K8s Frontend**

- Watch for `VolumeSnapshot` created for trident PVCs.
- Trigger a backend storage snapshot.
- On successful creation of a snapshot, create a `VolumeSnapshotContent` and bind it to the `VolumeSnapshot`. For example:
```
apiVersion: snapshot.storage.k8s.io/v1alpha1
kind: VolumeSnapshotContent
metadata:
  name: content-<snapshot-test-uid>
spec:
  csiVolumeSnapshotSource:
    driver: netapp.io/trident
    snapshotHandle: <backend-snapshot-id>
  volumeSnapshotRef:
    kind: VolumeSnapshot
    name: snapshot-test
    namespace: <snapshot-namespace>
    uid: <snapshot-test-UID>
  persistentVolumeRef:
    kind: "PersistentVolume"
    name: <pv-name>
    namespace: <pv-namespace>
    uid: <pv-UID>
```
This again is ignored by the `csi-snapshotter`.

- Update the `VolumeSnapshotStatus` with relevant information and `VolumeSnapshotSpec` with `snapshotContentName`.

**Trident Orchestrator**

- Create a snapshot on the corresponding backend storage.
- Store the state of the snapshot in the persistent store with `SnapshotID` as the key.

### CreateVolumeFromSnapshot flow

**User**

- Create a trident PVC with `dataSource` as the `VolumeSnapshot`
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: podpvc-from-snapshot
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: basic
  dataSource:
    kind: VolumeSnapshot
    name: snapshot-test
    apiGroup: snapshot.storage.k8s.io
```

**Trident K8s frontend**

- In `processPendingClaim`, similar to checking if a clone is required, check if a `dataSource` is present.
- Extract the `VolumeSnapshotContent` information from the `VolumeSnapshot` and then get the `snapshotHandle`.
- Pass on the `snapshotHandle` to create a volume from snapshot. 

**Trident Orchestrator**

- Fetch the snapshot object from the persistent store using the `snapshotHandle` (ID) as the key.
- Create a volume using the snapshot as the source.
The function signature would look like `CreateVolumeFromSnapshot(snapshotID string, volConfig *storage.VolumeConfig) (externalVol *storage.VolumeExternal, error)` 

### Some questions

- There are multiple ways to create the dummy `VolumeSnapshotClass`.
   - We can document it as a prerequisite for using the `VolumeSnapshot` feature and provide an example class for the user to create it manually.
   - We can also deploy a default class during the installation controlled by a flag.
What would you prefer?

- I was thinking of adding the code to install the CRDs during the Trident installation, probably with the install command. 
Should there be some flag to control this or is it okay to do it always?

Do let me know your thoughts on this.

**Comment by pavannd1** [comment]
Will reopen a fresh PR for this.

**Pull request closed without merging**